docs/source/format/CanonicalExtensions.rst

```diff
@@ -283,6 +283,28 @@ UUID
    A specific UUID version is not required or guaranteed. This extension represents
    UUIDs as FixedSizeBinary(16) with big-endian notation and does not interpret the bytes in any way.
 
+8-bit Boolean
+====
+
+Bool8 represents a boolean value using 1 byte (8 bits) to store each value instead of only 1 bit as in
+the native Arrow Boolean type. Although less compact that the native representation, Bool8 may have
+better zero-copy compatibility with various systems that also store booleans using 1 byte.
+
+* Extension name: ``arrow.bool8``.
+
+* The storage type of this extension is ``Int8`` where:
+
+  * **false** is denoted by the value ``0``.
+  * **true** can be specified using any non-zero value.
```

Comment:
```suggestion
  * **true** can be specified using any non-zero value. Preferably ``1``.
```

Reply:
It's the value of choice when casting booleans to integers on pretty much any system that allows this cast.

In C++, for instance, multiplication with `true` and `false` is a common pattern in branch-free programming.

```cpp
// not necessarily the best implementation of max, but it works.
// (it could overflow if a+b > MAX_INT, but the fact that C++
// compilers can assume + of signed ints doesn't overflows, allows
// it to detect this is just a max operation)
int branchfree_max(int a, int b) {
    bool condition = a > b;
    return a * condition + b * !condition;
}
```

This spec could also contain recommendations of what is expected when explicitly casting int arrays to 8-bit boolean extension type. The intX->8bit_bool cast SHOULD [1] canonicalize to 0 and 1 values.

C++ improved upon C's treatment of booleans by requiring that `bool` only range between `0` and `1` even though implicit casts are allowed.

```cpp
bool cast(int a) {
 return return a;
}
```

...becomes this in LLVM IR:

```llvm
  %tobool = icmp ne i32 %a, 0
  ret i1 %tobool
```

`icmp ne` means integer comparison (not equal) and `0` is the right-hand size operand. So any `int` other than `0` becomes `true`, but `true` and `false` tend to be `1` and `0` when a choice has to be made.

[1] SHOULD not MUST (https://www.rfc-editor.org/rfc/rfc2119). Users should watch out for values different of 1 and act accordingly.

Reply:
Thanks @felipecrv. Would these optimizations work if `1` is _preferred_ for `true` but any nonzero value is still considered valid? Perhaps there are some "fastpath" optimizations that can be done by checking the first bit (LE) first to see if the value is `1`, but any generic implementation can't use the absence of this bit to mean `false`. It will need to check all other bits as well. Let me know if I'm understanding your suggestion correctly.

> Producers MUST produce 0 or 1 values. Consumers SHOULD treat any non-zero value as true and 0 as false.

Can you clarify or give some examples of producers and consumers in this case? Unless I'm misunderstanding, if producers MUST produce 0 or 1 values then it's not clear how a consumer would ever receive any other value.

So I would then expect the inverse:
Producers **SHOULD** produce 0 or 1 values. Consumers **MUST** treat any non-zero value as true and 0 as false.

Reply:
I'm +1 with better using 0/1, but I'm also curious is there some system which would generate non 0-1 values? if not I think 0/1 is better

Reply:
>  Would these optimizations work if 1 is preferred for true but any nonzero value is still considered valid?

If the underlying memory for a `bool` value is not in {0, 1}, it's UB to multiply it.

In summary:
1) casting int to bool requires that the compiler normalize to `0` and `1`.
2) when using a `bool`, it's usually not an issue for it to be out of the [0,1] range. `if (b)` will work. `b * my_int` will UB. It's up to the user.

> Perhaps there are some "fastpath" optimizations that can be done by checking the first bit (LE) first to see if the value is 1

Checking that the whole byte is non-zero is a very fast operation. Faster than checking a single bit.

> Let me know if I'm understanding your suggestion correctly. [...] Unless I'm misunderstanding, if producers MUST produce 0 or 1 values then it's not clear how a consumer would ever receive any other value.

I think I articulated very badly my suggestion here. Let me try again in the next comment.

Reply:
When producing a `Bool8` array, integer values of any width MUST be normalized into the `[0, 1]` range of `int8` values. `0` maps to `0` and every other non-zero value maps to `1` (this is a very simple CPU operation).

A kernel consuming a `Bool8` array can assume values are in the `[0, 1]` range, but to be robust against less strictly-conformant producers, it's preferable to use `array[i] != 0` or `array[i] == 0` when converting the `int8` value to `bool` instead of comparing against `1`.

Compute kernels that would benefit from the `[0, 1]` assumption can rely on it and document that non-strictly conforming `Bool8` input may lead to bogus output.

Reply:
If someone finds a bug in a kernel because it assumed `[0, 1]` range and the `Bool8` array contains values out of the range, it's the producer that is violating the spec.

I believe systems that use bytes for bools are already normalized to `[0, 1]` so `MUST` should be a requirement on producers. It shouldn't not affect zero-copy.

And zero-copy goes both ways. Systems have both producers and consumers of arrays. So the way to order the requirements can never be perfect, but it can at least be compatible with the existing semantics adopted in programming languages and CPUs.

Reply:
my personal preference here is the wording that producers `SHOULD` produce 0 or 1 values, not `MUST` simply because there *might* be a system that uses non-0/1 values. Consumers shouldn't assume 0 and 1, they should only assume 0 and non-zero.

In practice, this will almost always be 0 and 1, but I'm generally in favor of fewer restrictions on an interoperable protocol where possible given that our stated goal is zero-copy. So supporting more systems is favorable. Since consumers can generally perform any and all operations based on just comparing against 0, I don't think we gain anything specifically by requiring producers to provide values in `[0, 1]`. 

While low-level C performance can be gained by multiplying with `true` and `false` for branch free programming, in the context of a compute kernel I think we shouldn't rely on that behavior. Most compute systems disallow multiplying bools anyways. Allowing zero-copy casts between Bool8 and Int8 without requiring a pass to check values and convert them is a good benefit I think.

Reply:
@felipecrv I think we're talking about mostly the same semantics but with slightly different phrasing.

The distinction between

**A**: "producers SHOULD produce 0 or 1 values"

and

**B**: "producers MUST produce 0 or 1 values" + affordances for "less strictly-conformant producers"

is very subtle. IMO the first statement is a simpler and clearer description of the specification.

I'll add 2 data points to the discussion to make things more concrete:
1. I did some [investigation](https://gist.github.com/joellubi/2ddf626633b57839cfd5f32cd94a7f3b) into how numpy handles this in the context of zero-copy sharing with pyarrow. It appears numpy does in fact canonicalize boolean values to 0 and 1, but understands any nonzero value to be true without forcing a copy. This aligns well with the behavior we're discussing.
2. libcudf defines its [BOOL8 type](https://docs.rapids.ai/api/libcudf/stable/group__utility__types#ggadf077607da617d1dadcc5417e2783539a05afd9eb8887a406d47474cd3809a5dd) as "Boolean using one byte per value, 0 == false, else true". It may be true that CUDF will often or even alway use the values 0 or 1 (I don't actually know), but it's not consistent with the documented behavior.

Reply:
>  I don't think we gain anything specifically by requiring producers to provide values in [0, 1].

@zeroshade We gain **the ability to say who is responsible for normalization** when there is a conflict in a bidirectional relationship (two systems that are both consumer and producer at the same time). This is what C++ had to do: int->bool normalizes while bool->int takes whatever is on the byte. Being loose on both directions is not the globally efficient solution because now both sides have to be paranoid instead of just one.

> While low-level C performance can be gained by multiplying with true and false for branch free programming, in the context of a compute kernel I think we shouldn't rely on that behavior. Most compute systems disallow multiplying bools anyways. 

Multiplying was just an example. Think about hashing/fingerprinting a `Bool8` array as another example where assuming normalized values is beneficial. An even more realistic example is comparing `Bool8` arrays for equality: if I implement `==(Bool8, Bool8)` as a simple fallback to `==(Int8, Int8)` (a very reasonable implementation) and it fails because arrays are not normalized I would ask the producers of the `Bool8` arrays to be more conservative in what they produce.

I believe this is how libcudf implements bool array comparison. This only works if producers of bool arrays stick to producing values in the `[0, 1]` range.

https://github.com/rapidsai/cudf/blob/34dea6fe40fc20966b48257853865111df4a687f/cpp/include/cudf/table/experimental/row_operators.cuh#L1293-L1297

At least one array producer in the codebase is producing arrays with values in the `[0, 1]` range — https://github.com/rapidsai/cudf/blob/branch-24.08/cpp/src/transform/mask_to_bools.cu#L55. The `bool` returned becomes `0` or `1` per CUDA/C++ conventions.

> Allowing zero-copy casts between Bool8 and Int8 without requiring a pass to check values and convert them is a good benefit I think.

A pass is not necessary because more often than not, the process that is producing 8-bit booleans is already producing 0s and 1s.

> Allowing zero-copy casts between Bool8 and Int8 without requiring a pass to check values and convert them is a good benefit I think.

All we need is the producer to have the equivalent of `reinterpret_cast<const bool*>(buffer)` to convert an `int8` array into a `Bool8` array without a pass that checks the values. This `reinterpret_cast` doesn't check byte by byte because the user is supposed to guarantee by construction of the buffer that the values on the `buffer` are in [0,1].

> @felipecrv I think we're talking about mostly the same semantics but with slightly different phrasing.

I think I'm being precious about a very nuanced topic (semantics + undefined behavior) and I'm not even able to convince others that this matters in the first place ;)

> 1. I did some [investigation](https://gist.github.com/joellubi/2ddf626633b57839cfd5f32cd94a7f3b) into how numpy handles this in the context of zero-copy sharing with pyarrow. It appears numpy does in fact canonicalize boolean values to 0 and 1, but understands any nonzero value to be true without forcing a copy. This aligns well with the behavior we're discussing.

This aligns with what I'm proposing. Normalize when you're producing and take non-zero to mean true.

> 2. libcudf defines its [BOOL8 type](https://docs.rapids.ai/api/libcudf/stable/group__utility__types#ggadf077607da617d1dadcc5417e2783539a05afd9eb8887a406d47474cd3809a5dd) as "Boolean using one byte per value, 0 == false, else true". It may be true that CUDF will often or even alway use the values 0 or 1 (I don't actually know), but it's not consistent with the documented behavior.

The libcudf spec doesn't contradicts anything I proposed, but it under-specifies the semantics (IMO) by not saying that, when a choice has to be made, `1` and not `0xff` or `8` is used to represent `true`. And if they did that (produced non-0 values non-deterministically to represent `true`) a lot would break within cudf itself (see the link to the `==` comparison above).

You don't have to say `MUST` in the spec. Just say that `1` is the value of choice when you're forced to pick a non-zero value to represent `true`.

> In practice, this will almost always be 0 and 1, but I'm generally in favor of fewer restrictions on an interoperable protocol where possible given that our stated goal is zero-copy.

I'm also going for flexibility here. I'm a fan of undefined-behavior and non-determinism as programming tools, but to maximize their effectiveness you have to say which side is responsible for making things deterministic when there is a conflict. While I say producers must pick `1` for true, I'm also hedging by saying that consumers should try to be robust against producers picking something other than `1`.

Reply:
This is too much text about a very nuanced topic and I'm sorry. It's not my intention to block the spec on this.

Reply:
I agree that it's a very nuanced topic. 

> I believe this is how libcudf implements bool array comparison. This only works if producers of bool arrays stick to producing values in the [0, 1] range.

> https://github.com/rapidsai/cudf/blob/34dea6fe40fc20966b48257853865111df4a687f/cpp/include/cudf/table/experimental/row_operators.cuh#L1293-L1297

Doesn't this fall into the C++ semantics we described earlier? For a bool array comparison, this will cast each value to bool to perform the comparisons which will, as per C++ semantics, do precisely what you mentioned earlier to perform the normalization as needed. Since C++ will perform the casting in this case, the only reason it would fail is if the compute did a memcmp or something rather than performing a comparison on each element.

> All we need is the producer to have the equivalent of reinterpret_cast<const bool*>(buffer) to convert an int8 array into a Bool8 array without a pass that checks the values. This reinterpret_cast doesn't check byte by byte because the user is supposed to guarantee by construction of the buffer that the values on the buffer are in [0,1].

I wasn't referring to the `reinterpret_cast` case where we know that the user has guaranteed it that the values are in [0, 1], I was referring to the row-based casting where each value is being cast to bool. If we state that producers of a `Bool8` array need to ensure that the values are in `[0, 1]`, then any cast implementation for int8 -> bool8 *MUST* at least perform a check to ensure the values are in [0, 1] if they intend on doing the equivalent of the `reinterpret_cast`. This is because the cast kernel itself didn't produce the int8 array, and thus can't know whether or not it conforms. Like you said, if we explicitly state that it *must* follow this, then one of the directions has to normalize. If we want consumers to be able to assume [0, 1] for Bool8, then any int8->bool8 cast will have to normalize instead of being able to do zero-copy (unless it does a pass to check first).

> You don't have to say MUST in the spec. Just say that 1 is the value of choice when you're forced to pick a non-zero value to represent true.

Isn't this equivalent to just saying *SHOULD*? If you agree we don't have to use `MUST`, then I think we're pretty much saying the same thing. The important distinction though, is that if we don't say that producers *MUST* use [0, 1], then consumers cannot and should not assume [0, 1]. Make sense?

I think at this point it makes more sense for us to summarize this and put it onto the mailing list and see if we can get more input from other people. I agree that this shouldn't hold up the spec though.

Reply:
> Doesn't this fall into the C++ semantics we described earlier? For a bool array comparison, this will cast each value to bool to perform the comparisons which will, as per C++ semantics, do precisely what you mentioned earlier to perform the normalization as needed. 

Not really. `clang++` compiles `lhs == rhs` to this assembly [1]:

```asm
Equals(bool, bool):                            # @Equals(bool, bool)
        mov     eax, edi
        xor     eax, esi
        xor     al, 1
        ret
```

Which is equivalent to `(lhs ^ rhs) ^ 1`. So if you call `Equals` [2] like this:

```cpp
bool Equals(bool lhs, bool rhs) {
    return lhs == rhs;
}

void VectorEquals(int n, const bool *lhs, const bool* rhs, bool *out) {
    for (int i = 0; i < n; i++) {
        // UBSAN builds will have instrumentation here to check that
        // the byte from memory being passed to a register as a "bool"
        // is in the valid range. Release builds don't care.
        out[i] = Equals(lhs[i], rhs[i]);
    }
}

void load_and_compare(int n) {
    // This cast can lead to UB if bytes are out of the [0, 1] range.
    auto *lhs = reinterpret_cast<const bool*>(LoadInt8Buffer(n));
    auto *rhs = reinterpret_cast<const bool*>(LoadInt8Buffer(n));
    auto *out = AllocBools(n);
    VectorEquals(n, lhs, rhs, out);
    PrintBools(n, out);
}
```

So let's say we load the `lhs` register with `2` (denoting `true`) and the `rhs` with `0` (denoting `false`). `Equals` of a value denoting `true` and a value denoting `false` should return a value denoting `false`, but:

```
(lhs ^ rhs) ^ 1 =
(10b ^ 00b) ^ 01b =
   10b      ^ 01b =
        11b  (3 evaluates to true!!)
```

[1] https://godbolt.org/z/3sM4debdT
[2] if libcudf is anything like Arrow code, that is how they would call the comparison function in a kernel loop

Reply:
>> You don't have to say MUST in the spec. Just say that 1 is the value of choice when you're forced to pick a non-zero value to represent true.
>
> Isn't this equivalent to just saying SHOULD? If you agree we don't have to use MUST, then I think we're pretty much saying the same thing. The important distinction though, is that if we don't say that producers MUST use [0, 1], then consumers cannot and should not assume [0, 1]. Make sense?

I'm saying to not use MUST/SHOULD language anymore because people do not necessarily have RFC2119 in mind when they read specs, but my intent, in non-RFC speak remains the same:

 - producers strive to produce only [0,1]-ranged bool8 values and if they don't, they are non-strictly conformant to the spec
 - consumers only assume [0,1] range if they understand that non-strictly compliant producers (a possibility in an imperfect world) might break their code. This is what SHOULD means in RFC2119.

The spec is a strict ideal. The implementations approximate the spec.

```
SHOULD This word, or the adjective "RECOMMENDED", mean that there
   may exist valid reasons in particular circumstances to ignore a
   particular item, but the full implications must be understood and
   carefully weighed before choosing a different course.
```

> I think at this point it makes more sense for us to summarize this and put it onto the mailing list and see if we can get more input from other people. I agree that this shouldn't hold up the spec though.

I thought this would be a non-controversial proposal:

```suggestion
  * **true** can be specified using any non-zero value. Preferably ``1``.
```

I don't feel like debating it too much anymore given that people naturally gravitate towards producing [0, 1]-ranged byte-sized bools no matter what the spec says.

It might be impossible to get consensus on a text that reconciles recommending the use of `1` while also saying that any non-zero value can be interpreted as `true`.

Reply:
Just for clarification on libcudf. The comparator referenced resolves the BOOL8 value to a C/C++ `bool` variable before calling the comparator and so follows the same semantics that a non-zero bool compares equal to another non-zero bool (nicely illustrated in the godbolt link here).
The magic is in the `column_device_view::element()` method which when called through with the type-dispatcher maps a `BOOL8` column type to a native `bool` type when reading the column row data from memory.
So libcudf does not require the bool8 values to be normalized to [0,1] but relies on the underlying behavior of C++ `bool` type.

Reply:
> I'm saying to not use MUST/SHOULD language anymore because people do not necessarily have RFC2119 in mind when they read specs, but my intent, in non-RFC speak remains the same:

That's fair and makes a lot of sense. I thought there was more to the intent of explicitly making it requirement for producers. Though I don't necessarily like the wording of "non-strictly compliant" in this context, we can recommend using `1` and say it is preferred (like we do in other areas such as offset buffers starting with 0) but ultimately the point is that we're explicitly allowing for it to not be in [0, 1] and therefore consumers can never assume it is unless they *know* the producer did that.

> I thought this would be a non-controversial proposal:

Honestly, I'm fine with that change given that we say it's recommended or that producers "should" produce [0, 1]. I don't think it's controversial at all.  @joellubi?

Reply:
> Honestly, I'm fine with that change given that we say it's recommended or that producers "should" produce [0, 1]. I don't think it's controversial at all. @joellubi?

I agree, and I think this discussion has surfaced valuable details about the systems we intend to enable interop with. I've committed the suggestion.